[AO] Add alert time range annotation for metric threshold

maryam-saeidi opened the pull request:
Resolves #153851

## Summary

This PR adds alert time range annotation for the metric threshold rule details page.

![image](https://user-images.githubusercontent.com/12370520/230080136-434bb586-c00e-4f36-aa11-525a38f8a650.png)
**Note**
I changed the rule criteria to stop the alert, hence the weird graph!

## 🧪 How to test
1. Add `xpack.observability.unsafe.alertDetails.metrics.enabled: true` to the Kibana config
2. Generate a metric threshold alert
3. Go to the related alert details page and check the annotation

maryam-saeidi committed:
Add alert time range annotation

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

maryam-saeidi committed:
Fix test

maryam-saeidi committed:
Merge branch 'main' into 153851-alert-time-range-annotation

elasticmachine commented:
Pinging @elastic/actionable-observability (Team: Actionable Observability)

maryam-saeidi committed:
Decrease opacity

maryam-saeidi committed:
Merge branch 'main' into 153851-alert-time-range-annotation

mohamedhamed-ahmed reviewed (approved):
LGTM!

maryam-saeidi committed:
Improve translation id

fkanout commented on x-pack/packages/observability/alert_details/src/components/alert_time_range_annotation.tsx:
```diff
@@ -0,0 +1,40 @@
+          coordinates: {
+            y0: 0,
+            x0: alertStart,
+            x1: alertEnd,
```
Why do we need to add the `x1`?

maryam-saeidi replied:
To only highlight the duration that the alert was firing.

fkanout replied:
The chart will stop showing data when the alert is ended. So the end of the chart is the end of the alert. This is how we did it in APM. 
Are you showing data after the alert ended in the Metric chart?

maryam-saeidi replied:
Yes, the calculation is similar to what we do for start: 20 mins after the alert was finished (unless `now` is before alertEnd + 20 mins)

Shall I merge my PR with the current implementation, and we discuss the end time of the chart with @maciejforcone  and @simianhacker ? (It should be an easy fix based on the decision)

simianhacker replied:
If the alert has ended, I agree we should do the same thing with the start and finish. This will come in handy when you're looking at a recovered alert; you'll be able to see the metric recover.

maryam-saeidi committed:
Improve translation id

maryam-saeidi committed:
Rename getAlertPaddedTimeRange to getPaddedAlertTimeRange

maryam-saeidi committed:
Merge branch 'main' into 153851-alert-time-range-annotation

fkanout commented:
@maryam-saeidi From the screenshot, I see that the red shade stopped, but the alert is still above the threshold. 🤔

maryam-saeidi commented:
> @maryam-saeidi From the screenshot, I see that the red shade stopped, but the alert is still above the threshold. 🤔

Yes, it is because of the note that I added beneath the chart:
>Note
>I changed the rule criteria to stop the alert, hence the weird graph!

fkanout reviewed (commented):
I followed the "How to test" section and tested the PR locally. I have an `active` alert with 10 minutes of the alert being triggered and +500 seconds of duration. I can't see the alert annotation or the alert newly added active red shade:
 
<img width="1193" alt="Screenshot 2023-04-05 at 17 35 16" src="https://user-images.githubusercontent.com/6838659/230131673-b4eaa3e6-b9eb-4066-91f4-61fbe8f7346b.png">

fkanout commented on x-pack/plugins/infra/public/alerting/metric_threshold/components/alert_details_app_section.tsx:
```diff
@@ -53,6 +60,12 @@ export function AlertDetailsAppSection({ alert, rule }: AppSectionProps) {
       dateFormat={uiSettings.get('dateFormat') || DEFAULT_DATE_FORMAT}
       id={ALERT_START_ANNOTATION_ID}
     />,
+    <AlertTimeRangeAnnotation
```
nit: 
I would name it with a mention of the _active_ alert status. e.g. `AlertActiveAreaAnnotations`

maryam-saeidi replied:
I didn't use `Area` in the name so that if we change the annotation from `rect annotation` to something else, we don't need to change the name.

I liked the `active` in the name of the component, so I rename it from `AlertTimeRangeAnnotation` to `AlertActiveTimeRangeAnnotation`

maryam-saeidi commented:
> I followed the "How to test" section and tested the PR locally. I have an `active` alert with 10 minutes of the alert being triggered and +500 seconds of duration. I can't see the alert annotation or the alert newly added active red shade:
> 
> <img alt="Screenshot 2023-04-05 at 17 35 16" width="1193" src="https://user-images.githubusercontent.com/6838659/230131673-b4eaa3e6-b9eb-4066-91f4-61fbe8f7346b.png">

Can you please share how I can reproduce it?

fkanout commented:
I re-bootstrapped Kibana and reran it; the issue is solved. It may be a package issue. Let's merge this one and keep an eye on the chart 
<img width="1187" alt="Screenshot 2023-04-05 at 18 01 34" src="https://user-images.githubusercontent.com/6838659/230138085-fe54e267-6eab-49d8-9bfa-44ec2d24ae4f.png">

fkanout commented on x-pack/packages/observability/alert_details/src/components/alert_time_range_annotation.tsx:
```diff
@@ -0,0 +1,43 @@
+          details: RECT_ANNOTATION_TITLE,
+        },
+      ]}
+      style={{ fill: color, opacity: 0.1 }}
```
I think we need to increase the opacity; it's not very visible Active shade vs Threshold shade 
<img width="529" alt="Screenshot 2023-04-05 at 18 09 28" src="https://user-images.githubusercontent.com/6838659/230139978-059474a8-e350-495f-96ee-40424a83944d.png">

maryam-saeidi replied:
I used the opacity that was mentioned in Figma, let's double check this one with @maciejforcone too.

fkanout reviewed (approved):
LGTM. Tested locally. I left some comments, but they are not blockers.

maryam-saeidi committed:
Rename AlertTimeRangeAnnotation to AlertActiveTimeRangeAnnotation

maryam-saeidi committed:
Update translation id

maryam-saeidi committed:
Merge branch 'main' into 153851-alert-time-range-annotation

CoenWarmer commented:
/oblt-deploy
